docs: add new doc on why Gatsby uses GraphQL (#11787)
gatsbyjs/gatsby: docs/docs/why-gatsby-uses-graphql.md

---
title: Why Gatsby Uses GraphQL
---

A common question about Gatsby is, “Why does Gatsby use GraphQL? Doesn’t it generate static files?”

Without providing some context, it can seem like GraphQL is overkill for something like Gatsby. In this document, you’ll see what problems arise when creating pages, and how those problems can be solved using GraphQL.

## Create a page without any data

For any kind of pages that aren’t directly created in `src/pages/`, you’ll need Gatsby’s [`createPages` Node API](/docs/node-apis/#createPages) to create pages programmatically.

All that’s required to create a page is a `path` where it should be created and the component that should be rendered there.

For example, if you had the following component:

```jsx:title=src/templates/no-data.js
import React from "react"

const NoData = () => (
  <section>
    <h1>This Page Was Created Programmatically</h1>
    <p>

>       No data was required to create this page — it’s just a React component!

Reviewer: This text is cut off in the code example. Not a huge deal, but I thought I'd pass that along!

Author: Cut off? Or scrolling?

Reviewer: I meant scrolling, sorry!

In the simplest cases, this is all that’s required for building pages with Gatsby. However, you’ll often want to pass data to the page so that the template component is reusable.

## Create a page with hard-coded data

To pass data to the created pages, you’ll need to pass `context` to the `createPage` call.

In `gatsby-node.js`, we can add context like so:

```js:title=gatsby-node.js
exports.createPages = ({ actions: { createPage } }) => {
  createPage({
    path: "/page-with-context/",
    component: require.resolve("./src/templates/with-context.js"),
    context: {
      title: "We Don’t Need No Stinkin’ GraphQL!",
      content: "<p>This is page content.</p><p>No GraphQL required!</p>",
    },
  })
}
```

The `context` property accepts an object, and we can pass in any data we want the page to be able to access.

When Gatsby creates pages, it includes a prop called `pageContext` and sets its value to `context`, so we can access any of the values in our component:

```jsx:title=src/templates/with-context.js
import React from "react"

const WithContext = ({ pageContext }) => (
  <section>
    <h1>{pageContext.title}</h1>

>     <div dangerouslySetInnerHTML={{ __html: pageContext.content }} />

Reviewer: A Senior Web Developer, international public speaker and winner of an O'Reilly Web Platform Award in 2016, Marcy is passionate about accessibility and test automation.

>     <div dangerouslySetInnerHTML={{ __html: pageContext.content }} />

Reviewer: Is there really no better standard than `dangerouslySetInnerHTML`? Can we recommend `ReactHtmlParser`, at least? https://www.npmjs.com/package/react-html-parser

Author: I'm hesitant to introduce another package at this point — I'd almost prefer to drop the HTML altogether before adding additional complexity like that.

We tend to expect that the headless CMS (or the user) is paying attention to innerHTML stuff, but I can see how that's probably making a few assumptions. 😅 

Does `react-html-parser` protect against any attack vectors? Or does it only avoid using `dangerouslySetInnerHTML`? I couldn't see anything about additional safety on the package page (I skimmed, though).

Reviewer: Understood about not wanting to add another package, and one that hasn't been thoroughly vetted. This is obviously a bigger discussion than this one post...so I'm happy to punt on it and discuss elsewhere. But it seems odd to me that the React docs say this: 
> So, you can set HTML directly from React, but you have to type out dangerouslySetInnerHTML and pass an object with a __html key, to remind yourself that it’s dangerous

And then we all use it anyway. ¯\_(ツ)_/¯

Author: I think it’s more just to remind people that we need to be sure we know what the HTML is. But I agree, it'd be great not to do that.

Reviewer 2: It's only dangerous for user generated input — when data comes from a CMS or other place you control about it's fine.

  </section>
)

export default WithContext
```

> Start the development server with `gatsby develop` and visit `localhost:8000/page-with-context/` to see the created page:

Author: Works for me. I'll update!

Once the JSON and the images are added, you can create product pages by importing the JSON into `gatsby-node.js` and loop through the entries to create pages:

```js:title=gatsby-node.js
exports.createPages = ({ actions: { createPage } }) => {
  const products = require("./data/products.json")
  products.forEach(product => {
    createPage({
      path: `/product/${product.slug}/`,
      component: require.resolve("./src/templates/product.js"),
      context: {
        title: product.title,
        description: product.description,
        image: product.image,
        price: product.price,
      },
    })
  })
}
```

The product template still uses `pageContext` to display the product data:

```jsx:title=src/templates/product.js
import React from "react"

const Product = ({ pageContext }) => (
  <div>
    <h1>{pageContext.title}</h1>
    <img
      src={pageContext.image}
      alt={pageContext.title}
      style={{ float: "left", marginRight: "1rem", width: 150 }}
    />
    <p>{pageContext.price}</p>
    <div dangerouslySetInnerHTML={{ __html: pageContext.description }} />
  </div>
)

export default Product
```

> Run `gatsby develop`, then open `localhost:8000/product/space-socks/` to see one of the generated products.

Author: As long as every starter supports this command, I'm on board with it. I'm going to do one big update for all instances (already ran the find-and-replace), so I'm not going to merge this suggestion, though.

> Run `gatsby develop`, then open `localhost:8000/product/space-socks/` to see one of the generated products.

Reviewer 3: > As long as every starter supports this command, I'm on board with it

They do! They also support `npm start` which is maybe a bit simpler, too.

> I'm going to do one big update for all instances (already ran the find-and-replace), so I'm not going to merge this suggestion, though.

I actually already did this! Some instances were intentionally left because they're referring to the command itself rather than _using_ the command. See gatsbyjs/gatsby#11556

> Run `gatsby develop`, then open `localhost:8000/product/space-socks/` to see one of the generated products.

Author: > I actually already did this!

I meant in this article only — just pushed that update. 😄

![Screenshot of a rendered product page.](./images/why-gql-product-json.png)

This gets the job done, but it has a few shortcomings that are going to get more complicated as time goes on:

1. The images and the product data are in different places in the source code.
2. The image paths are absolute from the _built_ site, not the source code, which makes it confusing to know how to find them from the JSON.
3. The images are unoptimized, and any optimization you do would have to be manual.
4. To create a preview listing of all products, we’d need to pass _all_ of the product info in `context`, which will get unweildy as the number of products increases.
5. It’s not very obvious where data is coming from in the templates that render the pages, so updating the data might be confusing later.

To overcome these limitations, Gatsby introduces GraphQL as a data management layer.

## Create pages using GraphQL

> There’s a bit more up-front setup required to get data into GraphQL, but the benefits often far outweigh the cost.

Reviewer 3: Maybe a bit stronger here? We can/should strongly encourage GraphQL.

```suggestion
There’s a bit more up-front setup required to get data into GraphQL, but the benefits far outweigh the cost.
```

Author: Yeah, I could go either way. I think there are definitely simple use cases where GraphQL isn’t needed, and we shouldn’t necessarily push folks toward GraphQL if they don't _need_ it. That being said, we can (and should) have a point of view, so I'm happy to make this a bit more strongly worded.

- Load the JSON file into GraphQL ([`gatsby-source-filesystem`](/packages/gatsby-source-filesystem/))
- Convert JSON files into GraphQL “nodes” ([`gatsby-transformer-json`](/packages/gatsby-transformer-json/))
- Optimize images ([`gatsby-plugin-sharp`](/packages/gatsby-plugin-sharp/))
- Convert images into GraphQL nodes ([`gatsby-transformer-sharp`](/packages/gatsby-transformer-sharp/))

In addition to the plugins, we’ll use [`gatsby-image`](/packages/gatsby-image/) to display the optimized images with lazy loading.

Install these packages using the command line:

```bash
npm install --save gatsby-source-filesystem gatsby-transformer-json gatsby-plugin-sharp gatsby-transformer-sharp gatsby-image
```

Then add them to `gatsby-config.js`:

```js:title=gatsby-config.js
module.exports = {
  plugins: [
    {
      resolve: "gatsby-source-filesystem",
      options: {
        path: "./data/",
      },
    },
    "gatsby-transformer-json",
    "gatsby-transformer-sharp",
    "gatsby-plugin-sharp",
  ],
}
```

To check that this worked, let’s use the GraphQL Playground, which is available during development, by running:

```
GATSBY_GRAPHQL_IDE=playground yarn develop
```

> > **NOTE:** The `GATSBY_GRAPHQL_IDE=playground` part of this command is optional. Adding it enables the GraphQL Playground instead of GraphiQL, which is an older interface for exploring GraphQL.

Reviewer 3: I would probably just have used the default here. It's nice to just document the happy path.

(But also I agree, I so much prefer the playground)

Reviewer 3: Also I'm aware this would require some updates to screenshots and other stuff, so this can totally stay too!

Author: (I actually did this because I thought the goal was to make the Playground the default after we’re sure it's not introducing any regressions and I wanted to avoid updating all the screenshots.) 😅

Reviewer 3: I like your style 😎 It's a lot better, in my opinion.

Here’s what that looks like in practice:

```jsx:title=src/templates/product-graphql.js
import React from "react"
import { graphql } from "gatsby"
import Image from "gatsby-image"

export const query = graphql`
  query($slug: String!) {
    productsJson(slug: { eq: $slug }) {
      title
      description
      price
      image {
        childImageSharp {
          fluid {
            ...GatsbyImageSharpFluid
          }
        }
      }
    }
  }
`

const Product = ({ data }) => {
  const product = data.productsJson

  return (
    <div>
      <h1>{product.title}</h1>
      <Image
        fluid={product.image.childImageSharp.fluid}
        alt={product.title}
        style={{ float: "left", marginRight: "1rem", width: 150 }}
      />
      <p>{product.price}</p>
      <div dangerouslySetInnerHTML={{ __html: product.description }} />
    </div>
  )
}

export default Product
```

A few notes about this file:

> 1. The result of the query is added to the template component as the `data` prop.

Reviewer 3: ```suggestion
1. The result of the query is injected to the template component as the `data` prop.
```

Author: I went back and forth on some of this wording. My intention is to avoid any computer science-ey phrasing to make sure we don't lose people on syntax. I know that's a losing battle, but here we are. 😄

Reviewer 3: Leave it :) It _feels like_ we generally use injected (so that's why I suggested) but maybe that term is in itself confusing?

Author: I think that it's one of those things that you don't know until you know. It's hard for me to be sure whether that's me being paranoid or if it's a valid concern, though.

2. The image path was automatically converted by the Sharp transformer into a “child node” that includes optimized versions of the image.
3. The query uses a [GraphQL fragment](https://www.gatsbyjs.org/packages/gatsby-image/#fragments) to query all the required data for optimized images. GraphQL fragments _do not work_ in the GraphQL Playground.
4. The `img` tag has been swapped out for a `gatsby-image` component named `Image`. Instead of a `src` attribute, it accepts an object with optimized image data.

Save this file, run `gatsby develop`, then open `localhost:8000/gql/purple-hat/`:

![Lazy loaded image of an angry cat wearing the purple hat.](./images/why-gql-images.gif)

The image is now optimized and lazy loaded.

After the initial setup, loading data with GraphQL is fairly similar to directly loading JSON, but it provides extra benefits like automatically optimizing images and keeping the data loading in the same place where it’s used.

> It’s certainly not required, and in simple use cases it might not be worth the effort, but in many sites and apps, GraphQL will vastly simplify the process of building and optimizing your pages.

Reviewer 3: I'd maybe again just tweak this to be stronger. We can and should hold that GraphQL is best practice for structuring Gatsby applications.

```suggestion
GraphQL is certainly not required--as shown earlier--but the benefits of adopting GraphQL are obvious and numerous. GraphQL will generally simplify the process of building and optimizing your pages, and so it is and will remain a best practice in structuring and writing your next, great Gatsby application.
```

Author: Made some tweaks here, but I like it!

> It’s certainly not required, and in simple use cases it might not be worth the effort, but in many sites and apps, GraphQL will vastly simplify the process of building and optimizing your pages.

Reviewer: I LOVE this post! It makes a great case for Gatsby and GraphQL, while introducing incremental steps with the added benefit of helping users accomplish different kinds of tasks (maybe context works for other small scenarios?). Nice work!

Author: Thanks!

exports.createPages = ({ actions: { createPage } }) => {
  createPage({
    path: "/no-data/",
    component: require.resolve("./src/templates/no-data.js"),
  })
}
```

After running `npm run develop`, you’ll see the following at `localhost:8000/no-data/`:

![Screenshot of the page generated by the previous code snippet.](./images/why-gql-no-data.png)

In the simplest cases, this is all that’s required for building pages with Gatsby. However, you’ll often want to pass data to the page so that the template component is reusable.

## Create a page with hard-coded data

To pass data to the created pages, you’ll need to pass `context` to the `createPage` call.

In `gatsby-node.js`, we can add context like so:

```js:title=gatsby-node.js
exports.createPages = ({ actions: { createPage } }) => {
  createPage({
    path: "/page-with-context/",
    component: require.resolve("./src/templates/with-context.js"),
    context: {
      title: "We Don’t Need No Stinkin’ GraphQL!",
      content: "<p>This is page content.</p><p>No GraphQL required!</p>",
    },
  })
}
```

The `context` property accepts an object, and we can pass in any data we want the page to be able to access.

> > **NOTE:** There are a few reserved names that _cannot_ be used in `context`. They are: `path`, `matchPath`, `component`, `componentChunkName`, `pluginCreator___NODE`, and `pluginCreatorId`.

Reviewer 2: do we document this anywhere else?

Author: Not that I'm aware of — the docs are pretty light on information about `context` in general. @pieh pointed me to the source code, which is where I grabbed these. https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/src/redux/actions.js#L134-L141

Reviewer: It could we good to document those elsewhere, it would work well on this page: https://www.gatsbyjs.org/docs/creating-and-modifying-pages/#pass-context-to-pages

I'll create an issue and PR for it.

5. It’s not very obvious where data is coming from in the templates that render the pages, so updating the data might be confusing later.

To overcome these limitations, Gatsby introduces GraphQL as a data management layer.

## Create pages using GraphQL

There’s a bit more up-front setup required to get data into GraphQL, but the benefits far outweigh the cost.

Using `data/products.json` as an example, by using GraphQL we’re able to solve all of the limitations from the previous section:

1. The images can be colocated with the products in `data/images/`.
2. Image paths in `data/products.json` can be relative to the JSON file.
3. Gatsby can automatically optimize images for faster loading and better user experience.
4. We no longer need to pass all product data into `context` when creating pages.
5. Data is loaded using GraphQL in the components where it’s used, making it much easier to see where data comes from and how to change it.

### Add the necessary plugins to load data into GraphQL

In order to load the product and image data into GraphQL, we need to add a few [Gatsby plugins](/plugins/). Namely, we need plugins to:

> - Load the JSON file into GraphQL ([`gatsby-source-filesystem`](/packages/gatsby-source-filesystem/))

Reviewer 4: I find the terminology a bit confusing here: what is a GraphQL node? What does it mean when data is loaded into GraphQL? Can we say something like "loaded into Gatsby's internal data store, that can be queried with GraphQL"?

Author: That makes sense, yeah. I'll clean it up.
